### PR TITLE
feat(4190): next.js reference store

### PR DESCRIPTION
**Pull request by alexcasche**
### WHY are these changes introduced?

Addresses [ENG-4198](https://nacelle.atlassian.net/browse/ENG-4198) 

- Builds a Next.js reference store using tailwind, and Contentful.

### WHAT is this pull request doing?

- Creates a Next.js flavored mirror image of our [Nuxt.js reference store](https://github.com/getnacelle/nacelle-js/pull/122)

## 

### How to Test

- `git checkout ENG-4190--next-reference-store`
- `cd references-stores/next && npm install`
- `npm run dev`

### CODE REVIEW

File: reference-stores/next/components/Header/HeaderMega.js [hunk 1]
```diff
@@ -0,0 +1,122 @@
+import { CSSTransition } from 'react-transition-group';
+import Link from 'next/link';
+import Image from 'next/Image';
```
Comment (jasonredick) [inline reference-stores/next/components/Header/HeaderMega.js:3]:
```suggestion
import Image from 'next/image';
```

File: reference-stores/next/README.md [hunk 2]
```diff
@@ -0,0 +1,19 @@
...
+
+# serve with hot reload at localhost:3000
+$ npm run dev
+
+# build for production and launch server
+$ npm run build
+$ npm run start
+
+For detailed explanation on how things work, check out the [Nacelle documentation](https://docs.getnacelle.com/nacellejs/next-starter.html#installation), and the
+[Nuxt documentation](https://nuxtjs.org).
```
Comment (jasonredick) [inline reference-stores/next/README.md:19]: Should be Next

File: reference-stores/next/components/Section/CtaStrip.js [hunk 3]
```diff
@@ -0,0 +1,53 @@
+import Image from 'next/Image';
```
Comment (jasonredick) [inline reference-stores/next/components/Section/CtaStrip.js:1]:
```suggestion
import Image from 'next/image';
```

File: reference-stores/next/components/Section/PromoStrip.js [hunk 4]
```diff
@@ -0,0 +1,43 @@
+import Image from 'next/Image';
```
Comment (jasonredick) [inline reference-stores/next/components/Section/PromoStrip.js:1]:
```suggestion
import Image from 'next/image';
```

File: reference-stores/next/components/Search/SearchAutocompleteItem.js [hunk 5]
```diff
@@ -0,0 +1,51 @@
...
+                <p className="flex flex-wrap justify-end ml-4">
+                  {price && (
+                    <span
+                      className={compareAtPrice ? 'text-red-600' : undefined}
+                    >
+                      {price}
+                    </span>
+                  )}
+                  {compareAtPrice && (
+                    <span class="ml-2 line-through">{compareAtPrice}</span>
```
Comment (jasonredick) [inline reference-stores/next/components/Search/SearchAutocompleteItem.js:39]:
```suggestion
                    <span className="ml-2 line-through">{compareAtPrice}</span>
```

File: reference-stores/next/context/Product.js [hunk 6]
```diff
@@ -0,0 +1,31 @@
+import { createContext, useState } from 'react'
+import { getSelectedVariant } from 'utils/getSelectedVariant'
+
+export const ProductContext = createContext({})
+
+export const ProductProvider = ({ children, product }) => {
+  const [selectedOptions, setSelectedOptions] = useState(product.variants[0].content.selectedOptions)
```
Comment (jasonredick) [inline reference-stores/next/context/Product.js:7]: Similar error as the Nuxt.js version: `TypeError: Cannot read property 'selectedOptions' of null` on /collections/all

**Review by jasonredick — CHANGES_REQUESTED** [review]
Great work! Similar issues as the Nuxt.js store:

- Overflowing autocomplete
Screen Shot 2022-05-04 at 4 45 59 PM

- No search filters
Screen Shot 2022-05-04 at 4 46 21 PM

File: reference-stores/next/components/Section/CtaStrip.js [hunk 7]
```diff
@@ -0,0 +1,53 @@
...
+import Link from 'next/Link';
```
Comment (NWRichmond) [inline reference-stores/next/components/Section/CtaStrip.js:2]: Casing:

```suggestion
import Link from 'next/link';
```

**Review by NWRichmond — CHANGES_REQUESTED** [review]
This has good bones! But there are a number of things I think we'll need to address before merging.

Most of the issues are fixed by #135.

Two other issues remain when testing locally:

1. Like in #122, the "Search All" button doesn't work when pressed, but the search page does load when hitting the `Enter` key.
2. For parity with the Nuxt Reference Store, products in search results should be wrapped in links to their respective PDPs.

**Review by NWRichmond — CHANGES_REQUESTED** [review]
Found some more instances of `className`~`s`~:

File: reference-stores/next/components/Header/HeaderMega.js (hunk 1 above)
Comment (NWRichmond) [reply reference-stores/next/components/Header/HeaderMega.js]: Fixed in https://github.com/getnacelle/nacelle-js/pull/133/commits/9d3d0c95de5016f1095d9151bebaddaf07f9e863.

File: reference-stores/next/components/Section/CtaStrip.js (hunk 3 above)
Comment (NWRichmond) [reply reference-stores/next/components/Section/CtaStrip.js]: Fixed in https://github.com/getnacelle/nacelle-js/pull/133/commits/9d3d0c95de5016f1095d9151bebaddaf07f9e863.

File: reference-stores/next/components/Section/PromoStrip.js (hunk 4 above)
Comment (NWRichmond) [reply reference-stores/next/components/Section/PromoStrip.js]: Fixed in https://github.com/getnacelle/nacelle-js/pull/133/commits/9d3d0c95de5016f1095d9151bebaddaf07f9e863.

File: reference-stores/next/.vscode/settings.json [hunk 8]
```diff
@@ -0,0 +1,10 @@
+{
+  "eslint.alwaysShowStatus": true,
+  "eslint.format.enable": true,
+  "editor.codeActionsOnSave": {
+    "source.fixAll.eslint": true
+  },
+  "editor.defaultFormatter": "esbenp.prettier-vscode",
+  "editor.formatOnSave": true,
+  "eslint.validate": ["javascript", "typescript", "vue"]
```
[the remark this answers is not in the document]
Comment (NWRichmond) [reply reference-stores/next/.vscode/settings.json]: Fixed in https://github.com/getnacelle/nacelle-js/pull/133/commits/9d3d0c95de5016f1095d9151bebaddaf07f9e863.

File: reference-stores/next/components/Collection/CollectionGrid.js [hunk 9]
```diff
@@ -0,0 +1,92 @@
...
+      const allProducts = [...products, ...collectionProducts]
+      setCanFetch(collectionProducts.length === 12)
+      setProducts(allProducts)
+      setActiveProducts(canFetch 
+        ? allProducts?.slice(0, allProducts.length - 1)
+        : allProducts
+      )
+    }
+    setIsFetching(false)
+  }
```
[the remark this answers is not in the document]
Comment (NWRichmond) [reply reference-stores/next/components/Collection/CollectionGrid.js]: Fixed in https://github.com/getnacelle/nacelle-js/pull/133/commits/9d3d0c95de5016f1095d9151bebaddaf07f9e863.

File: reference-stores/next/components/Section/CtaStrip.js (hunk 7 above)
Comment (NWRichmond) [reply reference-stores/next/components/Section/CtaStrip.js]: Fixed in https://github.com/getnacelle/nacelle-js/pull/133/commits/9d3d0c95de5016f1095d9151bebaddaf07f9e863.

File: reference-stores/next/components/Search/SearchForm.js [hunk 10]
```diff
@@ -0,0 +1,104 @@
...
+  const [resultsText, setResultsText] = useState()
+  const { 
+    query,
+    setQuery,
+    results
+  } = useSearch()
+
+  useEffect(() => {
+    setQuery({ query: router.query.q || '' })
+  }, [router.query.q])
```
[the remark this answers is not in the document]
Comment (NWRichmond) [reply reference-stores/next/components/Search/SearchForm.js]: Fixed in https://github.com/getnacelle/nacelle-js/pull/133/commits/9d3d0c95de5016f1095d9151bebaddaf07f9e863.

**Review by NWRichmond — CHANGES_REQUESTED** [review]
~Found some more instances of `className`~`s`~ (this time, tried to add the changes to a [batch](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/reviewing-changes-in-pull-requests/incorporating-feedback-in-your-pull-request)):~

_EDIT: it turns out `classNames` is part of the API for the [`<CSSTransition />` component](https://reactcommunity.org/react-transition-group/css-transition) from `'react-transition-group'`. Please disregard!_

**Comment by alexcasche** [comment]
Note to self - swap out `jsx` styles for css modules

**Comment by alexcasche** [comment]
Should be all set here now @NWRichmond @KrisQ @jasonredick @mdarrik

File: reference-stores/next/components/Search/SearchAutocomplete.js [hunk 11]
```diff
@@ -0,0 +1,70 @@
...
+          enterActive: styles.enterActive,
+          exitActive: styles.exitActive,
+          enter: styles.enter,
+          exit: styles.exit,
+          enterDone: styles.enterDone
+        }}
+      >
+        <div
+          className={`
+          w-full absolute md:right-0 md:w-96 md:top-10 bg-white overflow-hidden shadow rounded-lg
```
Comment (jasonredick) [inline reference-stores/next/components/Search/SearchAutocomplete.js:30]:
```suggestion
          w-full absolute lg:right-0 lg:w-96 lg:top-10 bg-white overflow-hidden shadow rounded-lg
```

Still noticed the width issue on tablet breakpoint, these changes should fix it.

**Review by NWRichmond — APPROVED** [review]
Excellent, thanks so much for those updates, @alexcasche! Things are working as expected across the homepage, PLPs, PDPs, Search, Cart, miscellaneous pages, and cart-to-checkout.

I'm really glad to see CSS Modules being used now - this should make for a much simpler scaffold of the forthcoming Gatsby Reference Store.

File: reference-stores/next/components/Product/ProductForm.js [hunk 12]
```diff
@@ -0,0 +1,153 @@
...
+    useProduct();
+
+  const { checkoutProcessing, addItem } = useCart();
+
+  const options =
+    product?.content?.options?.find((option) => {
+      return option.values.length > 1;
+    }) && product?.content?.options;
+
+  const price = formatPrice({ price: selectedVariant.price });
```
Comment (jasonredick) [inline reference-stores/next/components/Product/ProductForm.js:18]: I'm getting an error here on `npm run build`. I'm going to assume it's like the Nuxt issue where the variant is some how undefined. 

Screen Shot 2022-05-09 at 1 37 19 PM

**Review by jasonredick — CHANGES_REQUESTED** [review]
Still noticed the search width issue and proposed some changes to fix it. Also got an error during build that seems similar to the Nuxt variant undefined issue.

File: reference-stores/next/components/Search/SearchAutocomplete.js (hunk 11 above)
Comment (alexcasche) [reply reference-stores/next/components/Search/SearchAutocomplete.js:30]: @jasonredick caught this post-merge.  Will have a new small PR to fix shortly

**Pull request merged**